management_api_app/db/repositories/resource_specs.py

```diff
@@ -0,0 +1,37 @@
+from typing import List
+
+from azure.cosmos import ContainerProxy, CosmosClient
+
+from models.domain.resource import ResourceSpec
+from db.repositories.base import BaseRepository
+from core.config import STATE_STORE_RESOURCE_SPECS_CONTAINER
+
+
+class ResourceSpecRepository(BaseRepository):
+    def __init__(self, client: CosmosClient):
+        super().__init__(client, STATE_STORE_RESOURCE_SPECS_CONTAINER)
+
+    @property
+    def container(self) -> ContainerProxy:
+        return self._container
+
+    def _query(self, query: str):
+        resource_specs = list(self.container.query_items(query=query, enable_cross_partition_query=True))
+        return resource_specs
+
+    def get_by_name(self, name: str, latest=False) -> List[ResourceSpec]:
+        by_name = f"SELECT * FROM ResourceSpecs rs where rs.name = {name}"
+        if latest:
+            by_name += " AND rs.latest = true"
+        return self._query(by_name)
+
+    def get_by_name_and_version(self, name: str, version: str) -> ResourceSpec:
+        by_name_and_version = f"SELECT * FROM ResourceSpecs rs where rs.name = {name} AND rs.version = {version}"
+        specs = self._query(by_name_and_version)
+        assert(len(specs) == 1)  # name and version should have been unique
```

Comment:
There might not be a template with a given name/version - do we really want to assert? better throw an EntityNotFound error

Reply:
That is true. It should be
if specs:
   assert()
  return spec[0]
else:
  return None

I am not fond of throwing EntityNotFound in repository code. That should be done at controller level. If not found then result is None from the repo's perspective but if not found in controller it should be EntityNotFound and hence 404.

What do you think?

Reply:
but... that is literally what the EntityNotFound is for -- its not an HTTP exception - its a db error (see db.errors)